runtime/src/rad_claims.rs

```diff
@@ -0,0 +1,494 @@
+
+pub trait Trait: frame_system::Trait + pallet_balances::Trait {
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+    /// An expected duration of the session.
+    ///
+    /// This parameter is used to determine the longevity of `heartbeat` transaction
+    /// and a rough time when we should start considering sending heartbeats,
+    /// since the workers avoids sending them at the very beginning of the session, assuming
+    /// there is a chance the authority will produce a block and they won't be necessary.
+    type SessionDuration: Get<Self::BlockNumber>;
+
+    /// A configuration for base priority of unsigned transactions.
+    ///
+    /// This is exposed so that it can be tuned for particular runtime, when
+    /// multiple pallets send unsigned transactions.
+    type UnsignedPriority: Get<TransactionPriority>;
+
+    type AdminOrigin: EnsureOrigin<Self::Origin>;
+
+    type Currency: Currency<Self::AccountId>;
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as RadClaims {
+        /// Total unclaimed rewards for an account.
+        AccountBalances get(fn get_account_balance): map hasher(blake2_128_concat) T::AccountId => T::Balance = 0.into();
+        /// Map of root hashes that correspond to lists of RAD reward claim amounts per account.
+        RootHashes get(fn get_root_hash): map hasher(blake2_128_concat) T::Hash => bool;
+        /// Account that is allowed to upload new root hashes.
+        UploadAccount get(fn get_upload_account): T::AccountId;
+    }
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait>{
+        /// The combination of account id, amount, and proofs vector in a claim was invalid.
+        InvalidProofs,
+        /// The payout amount attempting to be claimed is less than the minimum allowed by [MIN_PAYOUT].
+        UnderMinPayout,
+        /// Amount being claimed is less than the available amount in [AccountBalances].
+        InsufficientBalance,
+        /// Protected operation, must be performed by admin
+        MustBeAdmin,
+    }
+}
+
+decl_event! {
+    pub enum Event<T> where
+        <T as frame_system::Trait>::AccountId,
+        <T as frame_system::Trait>::Hash,
+        <T as pallet_balances::Trait>::Balance,
+    {
+        Claimed(AccountId, Balance),
+        RootHashStored(Hash),
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event() = default;
+
+        /// Claims RAD tokens awarded through tinlake investments
+        /// Feeless dispatchable function
+        /// The extrinsic is validated by the custom `validate_unsigned` function below
+        ///
+        /// # <weight>
+        /// - Based on hashes length
+        /// # </weight>
+        #[weight = 195_000_000]
+        pub fn claim(origin,
+                     account_id: T::AccountId,
+                     amount: T::Balance,
+                     sorted_hashes: Vec<T::Hash>,
```

Comment:
I believe it is important to verify the proofs on each node. With the current flow, only the propagator node will verify that.

Reply:
Adding the check, wont harm anyways.